### PR TITLE
Dynamic and User Defined Nmap Stages

### PR DESCRIPTION
**Pull request by Ne0nd0g**
-Added default 'StagedNmapSettiings' for stage 0
		-This stage is used as a default if another stage is missing a setting
	-Added default 'stage[x]-flags' options for stages 0-5
	-Default sparta.conf file will include stage0 and stage[x]-flags settings
	-Dynamical build StagedNmapSettings based off the number of stages in the conf file
	-Modified "backupAndSave" to use dynamic nmap stages
	-Removed self.tools_nmap_stage[1-5]_ports variables
controller.py
	-modified "runCommand" function to dynamically count the total number of nmap stages
	-modified "runStagedNmap" command to dynamically work with stages from conf file
	-modified "runStagedNmap" command to work with user defined nmap flags
	-Verbose output for running staged nmap command
settingsdialogs.py
	**I Really didn't want to modify this, but some variables were tied in from the other modules**
	-Modified "updateSettinigs" to use dynamic nmap stages tools_nmap_stage1_ports
	-Modified "populateToolsTab to use dynamic nmap stages self.stage1Input getattr(self, str(stage) + "-input")
	-Modified "validateStageNmapTab" to use dynamic nmap stages
	-Added QtCore to imports
	-Added getStagedNmapSettings function to detect the number of stages in the .conf file

### CODE REVIEW

**Comment by Ne0nd0g** [comment]
Example user defined Nmap stages for sparta.conf:

```
[StagedNmapSettings]
stage0-ports=T:80
stage0-flags=-T4 -sV -sSU
stage1-ports=T:443
stage1-flags=-T4 -sV -sSU
stage2-ports="T:25,135,137,139,445,1433,3306,5432,U:137,161,162,1434"
stage2-flags=-T4 -sV -sSU -O -sC
stage3-ports="T:23,21,22,110,111,2049,3389,8080,U:500,5060"
stage4-flags=-T4 -sV -sSU
stage4-ports="T:0-20,24"
stage5-ports=T:80
stage5-flags=-T4 -sV -sSU
stage6-ports=T:22
stage6-flags=-T4 -sV -sSU
stage7-flags=-sV
stage8-ports=-
```

**Comment by st3r30byt3** [comment]
Thanks Russel. The idea is good and something we never got around to doing so thanks.
We are quite busy at the moment but whenever we get the chance we'll give it some good testing to ensure nothing else breaks and if that's the case, we'll include it in the next commit with due credit. ;)

**Comment by st3r30byt3** [comment]
This functionality will be in the next commit. Hopefully still in 2016 :)

**Comment by kildonan5** [comment]
its 2018 now. I dont see this commit in the latest release. Was this project abandoned, or was this commit forgotten?

File: controller/controller.py [hunk 1]
```diff
@@ -523,32 +523,30 @@ def runStagedNmap(self, iprange, discovery=True, stage=1, stop=False):
...
+					else:
+						command += "-p "+self.settings.stagedNmapSettings['stage'+str(stage)]['ports']
+				else:
+					command += "-p "+self.settings.stagedNmapSettings['stage0']['ports']
 			else:
-				command += "-sT "
-			command += "-p "+ports+' '+iprange+" -oA "+outputfile
-							
+				command += self.settings.stagedNmapSettings['stage'+str(stage)]['flags'] + ' ' + "-p " + \
+                           self.settings.stagedNmapSettings['stage0']['ports']
```
Comment (kildonan5) [inline controller/controller.py]: Is this part of the else block? I'm not much of a python programmer but the tabbing here is a little unclear to me and just wanted to confirm it was right

Comment (kildonan5) [reply controller/controller.py]: Disregard, I just realized its part of word wrap :P

**Comment by kildonan5** [comment]
As a suggestion (if this is not to be merged) consider outputting the nmap command being used so that at least the user knows what the staged scans are doing without having to dig into the python.

**Pull request closed without merging**